balancer-js/src/modules/pools/factory/composable-stable/composable-stable.factory.integration.spec.ts

```diff
@@ -0,0 +1,180 @@
+        [],
+        signer,
+        signerAddress,
+        to as string,
+        data as string
+      );
+      const composableStableFactoryInterface = new Interface(
+        composableStableFactoryAbi
+      );
+      const poolCreationEvent: LogDescription = findEventInReceiptLogs({
+        to: factoryAddress,
+        receipt: transactionReceipt,
+        logName: 'PoolCreated',
+        contractInterface: composableStableFactoryInterface,
+      });
+      if (poolCreationEvent) {
+        poolAddress = poolCreationEvent.args.pool;
+      }
+      expect(!!poolCreationEvent).to.be.true;
+      return;
+    });
+    it('should init join a pool', async () => {
+      const signerAddress = await signer.getAddress();
+      const composableStablePoolInterface = new Interface(
+        ComposableStablePoolAbi
+      );
+      const pool = new Contract(
+        poolAddress,
+        composableStablePoolInterface,
+        provider
+      );
+      const poolId = await pool.getPoolId();
+      const scalingFactors = await pool.getScalingFactors();
+      const amountsIn = [
+        parseFixed('10000', 6).toString(),
+        parseFixed('10000', 6).toString(),
+      ];
+      const { to, data } = composableStablePoolFactory.buildInitJoin({
+        joiner: signerAddress,
+        poolId,
+        poolAddress,
+        tokensIn: tokenAddresses,
+        amountsIn,
+      });
+      const { transactionReceipt, balanceDeltas } =
+        await sendTransactionGetBalances(
+          [...tokenAddresses, poolAddress],
+          signer,
+          signerAddress,
+          to,
+          data
+        );
+      const vaultInterface = new Interface(Vault__factory.abi);
+      const poolInitJoinEvent: LogDescription = findEventInReceiptLogs({
+        receipt: transactionReceipt,
+        to,
+        contractInterface: vaultInterface,
+        logName: 'PoolBalanceChanged',
+      });
+      expect(!!poolInitJoinEvent).to.be.true;
+      expect(
+        balanceDeltas
+          .slice(0, amountsIn.length)
+          .map((delta) => delta.toString())
+      ).deep.equal(amountsIn);
+
+      //Calculate and compare the bptAmountOut
+      const poolInvariant = StableMaths._invariant(
+        parseFixed(amplificationParameter, 3),
+        amountsIn.map((amount, index) => {
+          const upscaledAmount = _upscale(
+            BigInt(amount),
+            scalingFactors[index + 1].toBigInt()
+          ).toString();
+          return OldBigNumber(upscaledAmount, 10);
+        })
+      ).toString();
+      // The amountOut of BPT shall be (invariant - 10e6) for equal amountsIn
+      const bptAmountInvariantDelta = SolidityMaths.sub(
+        BigInt(poolInvariant),
+        BigInt(balanceDeltas[amountsIn.length].toString())
+      );
+      // 1e6 is the minimum bpt, this amount of token is sent to address 0 to prevent the Pool to ever be drained
+      expect(bptAmountInvariantDelta).to.equal(BigInt(1e6));
```

Comment:
Found this test a bit hard to reason against. How about something like:
```
      const expectedBptAmountOut = SolidityMaths.sub(
        BigInt(poolInvariant),
        BigInt(1e6)
      );
      expect(balanceDeltas[amountsIn.length].toBigInt()).eq(
        expectedBptAmountOut
      );
      ```
      (And keeping the useful comments)